[dagster-dbt] Refactor UnitTestDefinition instantiation in utils.py to use `from_dict` method

kang8 opened the pull request:
Fixed: https://github.com/dagster-io/dagster/issues/27007

## Summary & Motivation

Today I upgraded dbt to 1.9.1 but encountered an error while running dagster.
In the end, I found that UnitTestDefinition was not properly parsed.

In dbt, [this is also how they create UnitTestDefinition](https://github.com/dbt-labs/dbt-core/blob/8a8857a85c0cc66c7e3de9eb7e9ca7fd63d553a4/core/dbt/parser/unit_tests.py#L609).

Here are the details:

Updated the instantiation of UnitTestDefinition in the `__getattr__`
method to utilize the from_dict method instead of the constructor. This
change enhances the clarity and consistency of the code when handling
unit test nodes from the manifest JSON.


## How I Tested These Changes

Verified `test_dbt_with_unit_tests()` passes with both `fqn:*` and `tag:test` selectors.

https://github.com/dagster-io/dagster/blob/652af9acd0495eea8b66646cad2db5fce9079151/python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_asset_decorator.py#L1145-L1162

maximearmstrong commented:
Hey [kang8](https://github.com/kang8) - Thanks for your contribution!

I'm in favor of this change. What was the error message you received before? Would it possible to add a new unit test to make sure the use case is covered?

kang8 commented:
Just like issue #27007, when I upgrade dbt to version 1.9.1 and run `dagster dev`, an error occurs:

```
The above exception was caused by the following exception:
AttributeError: 'dict' object has no attribute 'enabled'

Stack Trace:
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_core/errors.py", line 287, in user_code_error_boundary
    yield
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_grpc/server.py", line 258, in __init__
    loadable_targets = get_loadable_targets(
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_grpc/utils.py", line 50, in get_loadable_targets
    else loadable_targets_from_python_module(module_name, working_directory)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_core/workspace/autodiscovery.py", line 32, in loadable_targets_from_python_module
    module = load_python_module(
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_core/code_pointer.py", line 134, in load_python_module
    return importlib.import_module(module_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/.local/share/uv/python/cpython-3.11.11-macos-aarch64-none/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/kang/self_dagster/dagster/__init__.py", line 4, in <module>
    from .assets import (
  File "/Users/kang/self_dagster/dagster/assets/__init__.py", line 2, in <module>
    from .customer_io import customer_io_assets
  File "/Users/kang/self_dagster/dagster/assets/customer_io.py", line 5, in <module>
    from ..assets.dbt import rightcapital_dbt_assets
  File "/Users/kang/self_dagster/dagster/assets/dbt.py", line 47, in <module>
    @dbt_assets(
     ^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster/_core/decorator_utils.py", line 223, in wrapped_with_context_manager_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster_dbt/asset_decorator.py", line 305, in dbt_assets
    specs, check_specs = build_dbt_specs(
                         ^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster_dbt/asset_utils.py", line 806, in build_dbt_specs
    selected_unique_ids = select_unique_ids_from_manifest(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dagster_dbt/utils.py", line 115, in select_unique_ids_from_manifest
    selector = graph_selector.NodeSelector(graph, manifest)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dbt/graph/selector.py", line 52, in __init__
    graph_members = {
                    ^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dbt/graph/selector.py", line 53, in <setcomp>
    unique_id for unique_id in self.full_graph.nodes() if self._is_graph_member(unique_id)
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kang/self_dagster/.venv/lib/python3.11/site-packages/dbt/graph/selector.py", line 177, in _is_graph_member
    return unit_test.config.enabled
           ^^^^^^^^^^^^^^^^^^^^^^^^

  warnings.warn(f"Error loading repository location {location_name}:{error.to_string()}")
```

Then I spent some time testing and found that there are currently no tests covering dbt 1.9. I have added tests for dbt 1.9 in 9452c84cb0.


PS: This is my first time learning about tox and buildkite, I'm not sure if I added the dbt 1.9 test correctly, if there are any issues, please feel free to point them out.

> Would it possible to add a new unit test to make sure the use case is covered?

Adding support for dbt 1.9 in the test is enough, because the existing tests have already covered this situation, it's just that the dbt 1.9 version was not tested before.

stiabh commented:
Any ETA on this fix? Thanks! @kang8 @maximearmstrong

maximearmstrong commented:
> Any ETA on this fix? Thanks! 

@kang8 @stiabh it will be included in next release
